File: PRACTICAS/practica1/Civitas_P1/src/civitas_p1/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;

import java.util.ArrayList;

/**
 *
 * @author xaviv
 */
public class CivitasJuego {
    
    private int indiceJugadorActual;
    private Tablero tablero;
    private MazoSorpresas mazo;
    private EstadosJuego estado;
    private GestorEstados gestorEstados;
    private ArrayList<Jugador> jugadores;
    
    public CivitasJuego(ArrayList<String> nombres){
        
        for(int i = 0; i< nombres.size();++i){
            
            Jugador jug= new Jugador(nombres.get(i));
            jugadores.add(jug);
            
        }
        gestorEstados = new GestorEstados();
        estado = gestorEstados.estadoInicial();
        indiceJugadorActual= Dado.getInstance().quienEmpieza(jugadores.size());
        mazo = new MazoSorpresas();
        
        
    }
    public boolean cancelarHipoteca(int ip){
        
        return jugadores.get(indiceJugadorActual).cancelarHipoteca(ip);
        
    }
    public boolean construirCasa(int ip){
        
        return jugadores.get(indiceJugadorActual).construirCasa(ip);
        
    }
    public boolean construirHotel(int ip){
        
        return jugadores.get(indiceJugadorActual).construirHotel(ip);
        
    }
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        
        int veces=tablero.getPorSalida();
        
        while(veces>0){
            jugadorActual.pasaPorSalida();
            --veces;
        }
        
    }
    public boolean finalDelJuego(){
        
        boolean fin = false;
        for(int i = 0; i < jugadores.size() && !fin;++i){
            
            if(jugadores.get(i).enBancarrota())
                fin=true;
        }
        return fin;
    }
    public Casilla getCasillaActual(){
        
        return tablero.getCasilla(jugadores.get(indiceJugadorActual).getNumCasillaActual());
        
    }
    public Jugador getJugadorActual(){
        
        return jugadores.get(indiceJugadorActual);
        
    }
    public boolean hipotecar(int ip){
        
        return jugadores.get(indiceJugadorActual).hipotecar(ip);
        
    }
    public String infoJugadorTexto(){
        
        return jugadores.get(indiceJugadorActual).toString();
        
    }
    private void inicializarMazoSorpresas(MazoSorpresas mazo){
        
        for(int i = 0; i < mazo.getsize(); ++i){
            
            this.mazo.alMazo(mazo.getSorpresa(i));
            
        }
                
    }
    private void inicializarTablero(Tablero tablero){
        
        this.tablero = new Tablero(tablero.getCarcel());
        int cont=0, i = 0;
        while(cont<2){
            
            Casilla cas = tablero.getCasilla(i);
            if(cas.getNombre()=="Salida" || cas.getNombre()=="salida"){
                cont++;
            }
            if(cont<2)
                this.tablero.añadeCasilla(cas);
            ++i;
        }
        
    }
    private void pasarTurno(){
        
        indiceJugadorActual = (indiceJugadorActual+1)%jugadores.size();
        
    }
    private ArrayList<Jugador> ranking(){
        int maximo, posicion=0;
        ArrayList<Jugador> ganadores= new ArrayList<Jugador>(jugadores.size());
        ArrayList<Jugador> auxiliar= new ArrayList<Jugador>(jugadores.size());
        for(int i = 0; i< jugadores.size();++i){
            
            Jugador jug= new Jugador(jugadores.get(i));
            ganadores.add(jug);
            
        }
        while(ganadores.size()>0){
            maximo=-3;
            for (int i = 0; i < ganadores.size() - 1; i++) {
                int cont = 0;
                for (int j = 0; j < ganadores.size() - i - 1; j++) {
                    cont = cont + ganadores.get(i).compareTo(ganadores.get(j));
                }
                if(cont>=maximo){

                    maximo=cont;
                    posicion=i;

                }
            }
            auxiliar.add(ganadores.get(posicion));
            ganadores.remove(ganadores.get(posicion));
        }
        
        return auxiliar;
    }
    public boolean salirCarcelPagando(){
        
        return jugadores.get(indiceJugadorActual).salirCarcelPagando();
        
    }
    public boolean salirCarcelTirando(){
        
        return jugadores.get(indiceJugadorActual).salirCarcelTirando();
        
    }
    public void siguientePasoCompletado(OperacionesJuego operacion){
        
        estado=gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual), estado, operacion);
        
    }
    public boolean vender(int ip){
        
        return jugadores.get(indiceJugadorActual).vender(ip);
        
    }
}

File: PRACTICAS/practica1/Civitas_Ruby/lib/civitas_juego.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas
class CivitasJuego
  
  private
    @indice_jugador_actual
    @jugadores
    @estado
    @gestor_estados
    @tablero
    @mazo
  
  def initialize(nombres)
   
    for i in (0..nombres.size)
            
      nu = Jugador.new(nombres[i])
      @jugadores.push(nu)
            
    end
        
    @estado = @gestor_estados.estado_inicial
    @indice_jugador_actual = Dado.instance.quien_empieza(@jugadores.size)
    @mazo = MazoSorpresas.new
    
  end


  def contabilizar_pasos_por_salida(jugador_actual)
        
    @por_salida = @tablero.get_por_salida
        
    while(@por_salida>0)
      jugador_actual.pasa_por_salida
      @por_salida-=1
    end
  end
  
  def final_del_juego()
        
    for i in (0..@jugadores.size)
      if(@jugadores[i].en_bancarrota)
        return true
            
    return false         
        
      end
    end
  end
  
  def get_casilla_actual()
        
       return @tablero.get_casilla(@jugadores[@indice_jugador_actual].num_casilla_actual)
  end

  def get_jugador_actual()
        
        return @jugadores[@indice_jugador_actual]
  end

  
  def info_jugador_texto()
        
    nuevo = Jugador.new(@jugadores[@indice_jugador_actual])
    info = "Nombre del jugador: #{nuevo.nombre} Esta en la casilla: #{get_casilla_actual} Tiene #{nuevo.get_propiedades.size}  
            propiedades y un saldo de: #{nuevo.saldo}" 
        
  if(nuevo.is_encarcelado)
    info += " Esta encarcelado " 
  else
    info += " No esta encarcelado " 
    
  end
  
  return info
                
  end      
  
  
  
  def inicializar_mazo_sorpresas(mazo)
        
    for i in(0..mazo.get_size)
      @mazo.al_mazo(mazo[i])
    end
  end
    
  def inicializar_tablero(tablero)
        
    @tablero = Tablero.new(tablero.carcel)
    for i in (0..tablero.size)
      @tablero.aniade_casilla(tablero[i])
        
    end
  end
  
  def pasar_turno()
        
        @indice_jugador_actual = @indice_jugador_actual +1 % @jugadores.size
        
  end
    
    
  def ranking()
        
    top = Array.new
    copiar = Array.new
        
    for i in (0..@jugadores.size)
      copiar.push(@jugadores[i])
    end  
        
    cnt = 0 
        
    while (cnt < @jugadores.size)
            
      pos = cnt;
            
      for j in (cnt+1..copiar.size)
                
        if(@jugadores[j].compare_to(@jugadores[j])== -1)
        pos = j 
                
        end
      end
            
      top.push(@jugadores[pos]);
      @jugadores.delete(@jugadores[pos]);
            
      cnt+=1
            
    end
        
        return top
  end
  


  def salir_carcel_pagando()
     
    return @jugadores[@indice_jugador_actual].salir_carcel_pagando
  end
    
  def salir_carcel_tirando()
    
    return @jugadores[indice_jugador_actual].salir_carcel_tirando
  end

  def siguiente_paso_completado(operacion)
     
    @estado = @gestor_estados.siguiente_estado(@jugadores[@indice_jugador_actual], @estado, operacion)
        
  end
    
  def vender(ip)
        
    return @jugadores[@indice_jugador_actual].vender(ip)
        
  end





































end
end